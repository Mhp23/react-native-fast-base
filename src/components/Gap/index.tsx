import React from 'react';
import Divider from '../Divider';
import {useRM} from 'react-native-full-responsive';
import {DefaultSizes, GapProps, PropsWithChildren} from '../../core';

const DEFAULT_SPACE = 5;

const Gap: React.FC<PropsWithChildren<GapProps>> = ({
  children,
  mode = 'V',
  space = 'xs',
  dividerProps,
  reversed = false,
  behavior = 'divider',
}) => {
  const {rs} = useRM();

  const toArray = React.Children.toArray(children);

  const childrens = reversed ? toArray.reverse() : toArray;

  return (
    <React.Fragment>
      {childrens.map((child, index) => {
        if (React.isValidElement(child)) {
          const key = child?.key ?? `gap-i-${index}`;
          const isThreshold = index === toArray?.length - 1;
          const spaceValue =
            (typeof space === 'number' ? space : DefaultSizes[space]) ||
            DEFAULT_SPACE;
          if (behavior === 'style' && child.type !== React.Fragment) {
            const responsiveValue = rs(spaceValue);
            const style = isThreshold
              ? {}
              : mode === 'H'
              ? {marginEnd: responsiveValue}
              : {marginBottom: responsiveValue};
            return React.cloneElement<any>(child, {
              key,
              style: {...child.props.style, ...style},
            });
          } else {
            const DividerComponent = (
              <Divider
                mode="padding"
                width={spaceValue / 2}
                direction={mode === 'H' ? 'horizontal' : 'vertical'}
                {...dividerProps}
              />
            );
            return (
              <React.Fragment key={key}>
                {child}
                {!isThreshold && DividerComponent}
              </React.Fragment>
            );
          }
        } else {
          return child;
        }
      })}
    </React.Fragment>
  );
};

export default React.memo(Gap);
