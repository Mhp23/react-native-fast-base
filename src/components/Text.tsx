import React from 'react';
import {TextProps} from '../core/types';
import {
  StyleProp,
  Text as NativeText,
  TextStyle,
  TouchableOpacity,
} from 'react-native';
import {responsiveFontSize} from 'react-native-responsive-dimensions';
import {useTheme} from '../hooks';

const defaultProps: TextProps = {
  alignX: 'auto',
};

const Text: React.FC<TextProps> = ({
  children,
  lineHeight,
  alignX,
  style,
  color,
  font,
  size,
  onPress,
  ...rest
}) => {
  const {colors} = useTheme();

  const textStyles: StyleProp<TextStyle> = {
    fontFamily: font,
    fontSize: responsiveFontSize(size || 1.5),
    color: color ? color : colors?.text,
    textAlign: alignX,
    lineHeight: lineHeight,
  };

  return onPress ? (
    <TouchableOpacity activeOpacity={0.5} onPress={onPress}>
      <NativeText style={[textStyles, style]} {...rest}>
        {children}
      </NativeText>
    </TouchableOpacity>
  ) : (
    <NativeText style={[textStyles, style]} {...rest}>
      {children}
    </NativeText>
  );
};

Text.defaultProps = defaultProps;

export default Text;
