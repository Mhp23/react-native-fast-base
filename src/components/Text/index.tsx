import React from 'react';
import {useStyle, useTheme} from '../../hooks';
import {useRM} from 'react-native-full-responsive';
import {colorSelector, makeStyle} from '../../utils';
import {DefaultTextSizes, type TextProps} from '../../core';
import {StyleSheet, Text as NativeText, type TextStyle} from 'react-native';

const defaultProps: TextProps = {
  size: 'md',
};
/**
 * in typescript, the Text component allows you to use custom fonts that
 * you added to your project, only need to wrap your type inside it like
 * the below example:
 * ```
 * <Text<'MyFont1' | 'MyFont2'> {...props}>
 *  Awesome Text!
 * </Text>
 * ```
 */
const Text = React.forwardRef(
  <T extends string = ''>(
    {
      ax,
      font,
      size,
      color,
      style,
      weight,
      height,
      children,
      ...rest
    }: TextProps<T>,
    ref?: React.Ref<NativeText>,
  ) => {
    const {colors} = useTheme();

    const {rs} = useRM();

    const textStyles = useStyle<TextStyle>(() => {
      const textColor = colorSelector(color) || colors?.text;
      const fontSize = rs(
        typeof size === 'number' ? size : DefaultTextSizes[size],
      );
      return makeStyle({
        fontSize,
        textAlign: ax,
        fontFamily: font,
        color: textColor,
        fontWeight: weight,
        lineHeight: height,
      });
    }, [color, colors?.text, rs, size, font, ax, weight, height]);

    return (
      <NativeText
        ref={ref}
        testID="FAST_BASE_TEXT"
        style={StyleSheet.flatten([textStyles, style])}
        {...rest}>
        {children}
      </NativeText>
    );
  },
);

Text.defaultProps = defaultProps;

export default Text as <T extends string = ''>(
  props: TextProps<T> & {ref?: React.Ref<NativeText>},
) => React.ReactElement;
