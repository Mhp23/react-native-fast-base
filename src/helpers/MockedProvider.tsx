import React from 'react';
import {DefaultTheme, ThemeContext, type ThemeContextProps} from '../core';

const MockedProvider: React.FC = ({children}) => {
  return (
    <ThemeContext.Provider
      value={{
        ...({} as ThemeContextProps),
        ...DefaultTheme.DefaultTheme,
      }}>
      {children}
    </ThemeContext.Provider>
  );
};

export default MockedProvider;
