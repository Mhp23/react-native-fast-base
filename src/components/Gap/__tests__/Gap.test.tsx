import React from 'react';
import Gap from '..';
import Text from '../../../components/Text';
import MockedProvider from '../../../helpers/MockedProvider';
import {render, cleanup} from '@testing-library/react-native';

afterEach(cleanup);

describe('Gap component', () => {
  it('should render correctly when behavior is the divider', () => {
    const rendered = render(
      <MockedProvider>
        <Gap behavior="divider" space="md">
          <Text>Hello World</Text>
          <Text size="lg">Hello World</Text>
          <Text size="xl">Hello World</Text>
        </Gap>
      </MockedProvider>,
    );
    expect(rendered).toBeTruthy();
    expect(rendered).toMatchSnapshot();
  });

  it('should render correctly when behavior is the style', () => {
    const rendered = render(
      <MockedProvider>
        <Gap behavior="style" space="md">
          <Text>Hello World</Text>
          <Text size="lg">Hello World</Text>
          <Text size="xl">Hello World</Text>
        </Gap>
      </MockedProvider>,
    );
    expect(rendered).toBeTruthy();
    expect(rendered).toMatchSnapshot();
  });

  it('should render the fragment parented components correctly', () => {
    const rendered = render(
      <MockedProvider>
        <Gap space="md">
          <React.Fragment>
            <Text>Hello World</Text>
          </React.Fragment>
          <Text size="xl">Hello World</Text>
          <>
            <Text size="lg">Hello World</Text>
          </>
        </Gap>
      </MockedProvider>,
    );
    expect(rendered).toBeTruthy();
    expect(rendered).toMatchSnapshot();
  });
});
