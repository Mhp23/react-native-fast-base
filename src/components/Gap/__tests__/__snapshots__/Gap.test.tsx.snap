// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gap component should render correctly when behavior is the divider 1`] = `
Array [
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 33.333333333333336,
      }
    }
    testID="FAST_BASE_TEXT"
  >
    Hello World
  </Text>,
  <View
    style={
      Object {
        "paddingVertical": 12.5,
      }
    }
    testID="FAST_BASE_DIVIDER"
  />,
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 37.5,
      }
    }
    testID="FAST_BASE_TEXT"
  >
    Hello World
  </Text>,
  <View
    style={
      Object {
        "paddingVertical": 12.5,
      }
    }
    testID="FAST_BASE_DIVIDER"
  />,
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 41.66666666666667,
      }
    }
    testID="FAST_BASE_TEXT"
  >
    Hello World
  </Text>,
]
`;

exports[`Gap component should render correctly when behavior is the style 1`] = `
Array [
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 33.333333333333336,
        "marginBottom": 25,
      }
    }
    testID="FAST_BASE_TEXT"
  >
    Hello World
  </Text>,
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 37.5,
        "marginBottom": 25,
      }
    }
    testID="FAST_BASE_TEXT"
  >
    Hello World
  </Text>,
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 41.66666666666667,
      }
    }
    testID="FAST_BASE_TEXT"
  >
    Hello World
  </Text>,
]
`;

exports[`Gap component should render the fragment parented components correctly 1`] = `
Array [
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 33.333333333333336,
      }
    }
    testID="FAST_BASE_TEXT"
  >
    Hello World
  </Text>,
  <View
    style={
      Object {
        "paddingVertical": 12.5,
      }
    }
    testID="FAST_BASE_DIVIDER"
  />,
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 41.66666666666667,
      }
    }
    testID="FAST_BASE_TEXT"
  >
    Hello World
  </Text>,
  <View
    style={
      Object {
        "paddingVertical": 12.5,
      }
    }
    testID="FAST_BASE_DIVIDER"
  />,
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 37.5,
      }
    }
    testID="FAST_BASE_TEXT"
  >
    Hello World
  </Text>,
]
`;
